clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -711,13 +711,11 @@ private int partition(ProducerRecord<K, V> record, byte[] serializedKey , byte[]
         Integer partition = record.partition();
         if (partition != null) {
             List<PartitionInfo> partitions = cluster.partitionsForTopic(record.topic());
-            int numPartitions = partitions.size();
+            int numPartitions = partitions.size() - 1;
```

Comment:
The actual purpose of this change was to use the same value not in the check, but also in IllegalArgumentException instantiation.

So two options are possible:
1. Either use "numPartitions - 1" in both places.
2. Or rename the variable to "lastPartition".

I'd prefer the option 2.